### PR TITLE
Add the ability to patch an existing resource's JSON by including diff

### PR DESCRIPTION
**Pull request by alexbartlow**
Also add a Readme.md, with instructions on how to develop against ModTek.

### CODE REVIEW

**Comment by CptMoore** [comment]
this will only work for simple changes right? how to change lists? (add a missle hardpoint to the left torso of a mech).
I think you need to allow the use of a json query language to find data and allow modders to use that to replace data.



alternatively use dynamic compiling C# code and your patch hooks, this could already be enough for modders to change data.

```
File: mod.cs
void jsonResourceLoadHook(JSONObject json) {
 if (json["Description"]["Id"] == "Weapon_Laser_MediumLaser_1-Magna") {
  json["Damage"] = 20;
 }
}
```

https://stackoverflow.com/questions/826398/is-it-possible-to-dynamically-compile-and-execute-c-sharp-code-fragments

**Comment by Mpstark** [comment]
I've merged in some changes from master and renamed a couple things to my liking. The largest change in functionality that I changed was to include the ability for a `ModDef.ManifestEntry` to be able to specify that it should/shouldn't be patched, with the default being true.

I'll play with it, we can discuss it further, but the plan is to merge it soon.

**Comment by Mpstark** [comment]
@CptMoore, I'll do some digging before we commit to this, obviously, but this is a core feature that I want in the initial release. We'll just have to document how it works and any other behavior can be tacked on later.

**Comment by CptMoore** [comment]
looks alright based on the things you said except for the ManifestEntry merging, where do I now specify what should or should not be merged?

**Comment by Mpstark** [comment]
Setting it in the manifest with `ModDef.ManifestEntry.ShouldMergeJSON`. There is no implicit loading yet.

**Comment by Mpstark** [comment]
@alexbartlow If you've got time, I'd like a last sanity check before merging -- I'd like to merge today.

File: ModTek/Patches.cs [hunk 1]
```diff
@@ -29,11 +80,30 @@ static void Postfix(VersionManifest __result)
...
-                var type = entryKVP.Value.Type;
+                var newEntry = entryKVP.Value;
+                
+                if (newEntry.ShouldMergeJSON && __result.Contains(id, newEntry.Type))
+                {
+                    // read the manifest pointed entry and hash the contents
+                    DoJSONMerge.JSONHashes.Add(File.ReadAllText(__result.Get(id, newEntry.Type).FilePath).GetHashCode());
+
+                    // The manifest already contains this information, so we need to queue it to be merged
+                    var partialJSON = File.ReadAllText(newEntry.Path);
```
Comment (alexbartlow) [inline ModTek/Patches.cs]: Is this reading the file twice?

**Comment by alexbartlow** [comment]
@Mpstark I'm assuming that you've tested it with these changes, but it looks good from a code-review perspective. I just had one comment about a place where it looks like we're reading the JSON twice.

File: ModTek/Patches.cs (hunk 1 above)
Comment (Mpstark) [reply ModTek/Patches.cs]: No, that's reading two separate files, the source file that we're merging onto and the merge itself.

**Pull request merged**